src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -0,0 +1,315 @@
+    private lateinit var chid: String
+
+    private lateinit var role: AtomicInteger
+
+    private val handshakeHandlerName = "NoiseHandshake"
+
+    override val announce = Companion.announce
+    override val matcher = ProtocolMatcher(Mode.PREFIX, name = "/noise/$protocolName/0.1.0")
+
+    init {
+        logger = LogManager.getLogger(loggerNameParent)
+        Configurator.setLevel(loggerNameParent, Level.DEBUG)
+    }
+
+    // simplified constructor
+    fun initChannel(ch: P2PAbstractChannel): CompletableFuture<SecureChannel.Session> {
+        return initChannel(ch, "")
+    }
+
+    override fun initChannel(
+            ch: P2PAbstractChannel,
+            selectedProtocol: String
+    ): CompletableFuture<SecureChannel.Session> {
+        role = if (ch.isInitiator) AtomicInteger(HandshakeState.INITIATOR) else AtomicInteger(HandshakeState.RESPONDER)
+
+        chid = "ch=" + ch.nettyChannel.id().asShortText() + "-" + ch.nettyChannel.localAddress() + "-" + ch.nettyChannel.remoteAddress()
+        logger.debug(chid)
+
+        val ret = CompletableFuture<SecureChannel.Session>()
+        val resultHandler = object : ChannelInboundHandlerAdapter() {
+            override fun userEventTriggered(ctx: ChannelHandlerContext, evt: Any) {
+                when (evt) {
+                    is SecureChannelInitialized -> {
+                        val session = evt.session as NoiseSecureChannelSession
+                        ctx.channel().attr(SECURE_SESSION).set(session)
+
+                        ret.complete(session)
+                        ctx.pipeline().remove(this)
+                        ctx.pipeline().addLast(NoiseXXCodec(session.aliceCipher, session.bobCipher))
+
+                        logger.debug("Reporting secure channel initialized")
+                    }
+                    is SecureChannelFailed -> {
+                        ret.completeExceptionally(evt.exception)
+
+                        ctx.pipeline().remove(handshakeHandlerName + chid)
+                        ctx.pipeline().remove(this)
+
+                        logger.debug("Reporting secure channel failed")
+                    }
+                }
+                ctx.fireUserEventTriggered(evt)
+                ctx.fireChannelActive()
+            }
+        }
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName + chid, NoiseIoHandshake())
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName + chid + "ResultHandler", resultHandler)
+        return ret
+    }
+
+    inner class NoiseIoHandshake() : SimpleChannelInboundHandler<ByteBuf>() {
+
+        private val handshakestate: HandshakeState = HandshakeState(protocolName, role.get())
+        private var loggerName: String
+        private var logger2: Logger
+
+        private var localNoiseState: DHState
+        private var sentNoiseKeyPayload = false
+
+        private val instancePayload = ByteArray(65535)
```

Comment:
Can we avoid creating max-length buffer here?

Reply:
Avoiding max-length buffers requires dynamic bytearrays since the size isn't known beforehand, which will require modifying the Noise library that's been brought in. I'll take a look how to do that.

Reply:
Done. The length is now as tight as the longest buffer needed during the handshake (80 extra bytes in the worst case), and it doesn't limit adding arbitrary payloads at any step.

Reply:
Thanks! Cause e.g. having 100K simultaneous  connections would else require reserving 6G of useless memory